### PR TITLE
Eliminate post groups

### LINKED ISSUE #2666: Eliminate post groups
### What new functionality do you need?

UPP and products tasks output times are currently grouped to process multiple hours in one task. However, this greatly complicates downstream dependencies and we now favor clarity over reducing the number of tasks.

Prerequisite to #1251 

### What are the requirements for the new functionality?

Eliminate the usage of groups for UPP and products tasks. Instead, each task should handle a single output time.

### Acceptance Criteria

- [ ] Each upp and product task does one forecast hour without a group
- [ ] Output identical to existing output

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by WalterKolczynski-NOAA**
# Description
Eliminates the post groups used for upp and products jobs so that each task only processes one forecast hour. This is more efficient and greatly simplifies downstream dependencies that depend on a specific forecast hour.

Resolves #2666
Refs #2642

# Type of change
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Cycled test on Hercules

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA I'm just curious how this effects things like really long runs - say SFS length runs.  The grouping of the post jobs means way less post jobs, which might be nice for some systems.  I do appreciate the simplification though.  Just wanted to put this in the discussion since I recently ran a long run and well, it was a lot of post jobs!!!

**Comment by aerorahul** [comment]
More or less (post) jobs translate into longer or shorter walltime requests.  Theoretically, breaking the group into smaller requests (walltime, etc), gets through the batch scheduler faster. It should also help in debugging/restarting failed jobs without rerunning successful parts of the failed group.
Collecting the jobs in groups is the developer's way of gaming the scheduler.  Its appealing, but someone needs to show objectively that having fewer jobs with longer wall times is faster than more jobs with shorter turn around.
One could put all the jobs in a single job request and churn through as many cycles, tasks as possible, resulting in a single job, but thats an extreme (it has been seen!) This would eliminate the need for a workflow or workflow manager.
It will also reduce the number of jobs in total.  
The number of jobs in the workflow is perhaps not something we should worry about, but how they are executed and managed/driven.

Considering that this workflow will work on big and small machines, as well as on the cloud, this change manages resources better.

That's my 2c.

File: jobs/rocoto/upp.sh [hunk 1]
```diff
@@ -44,16 +44,11 @@ export PYTHONPATH
 export job="upp"
 export jobid="${job}.$$"
 
-###############################################################
-# shellcheck disable=SC2153,SC2001
-IFS='_' read -ra fhrs <<< "${FHRLST//f}" # strip off the 'f's convert to array
+export FORECAST_HOUR=$(( 10#${FHR3} ))
```
Comment (aerorahul) [inline jobs/rocoto/upp.sh:47]: The UPP also processes analysis for which FHR3 is `000` (This is needed in upp.x).
Just a note why there isn't the same logic as ATMOS_PRODUCTS here to handle the negative `-001`

**Review by aerorahul — APPROVED** [review]
looks good to me.

**Review by KateFriedman-NOAA — APPROVED** [review]
Looks good, thanks @WalterKolczynski-NOAA !

**Comment by GwenChen-NOAA** [comment]
@WalterKolczynski-NOAA, I'm working on breaking the gempak job from running all forecast hours at once (more than one hour runtime) to processing one forecast hour at a time, same as the awips job. The new scripts run successfully on WCOSS2 using my driver script, and it reduces the runtime to less than 4 min per task. It requires some workflow changes to accommodate the new scripts. Can those changes be made in this PR?

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA, I'm working on breaking the gempak job from running all forecast hours at once (more than one hour runtime) to processing one forecast hour at a time, same as the awips job. The new scripts run successfully on WCOSS2 using my driver script, and it reduces the runtime to less than 4 min per task. It requires some workflow changes to accommodate the new scripts. Can those changes be made in this PR?

They should be there own PR, but this PR should facilitate that work by making the prerequisites easier (the gempak job will only have to wait for the products job of the same forecast hour, instead of all of them). I'm helping Bo do the same thing for the sounding job. I'm on leave most of today, but send me an email reminder and we can coordinate early next week.

**Comment by GwenChen-NOAA** [comment]
> They should be there own PR, but this PR should facilitate that work by making the prerequisites easier (the gempak job will only have to wait for the products job of the same forecast hour, instead of all of them). I'm helping Bo do the same thing for the sounding job. I'm on leave most of today, but send me an email reminder and we can coordinate early next week.

Thanks, @WalterKolczynski-NOAA! I will open a new PR to upload my new gempak scripts, so you can take a look. We can coordinate next week.

**Comment by TerrenceMcGuinness-NOAA** [comment]
Timed out on Orion:
https://github.com/emcbot/ci-global-workflows/tree/error_logs/ci/error_logs/timed_out

**Comment by WalterKolczynski-NOAA** [comment]
The wrong forecast hour lists are being generated for marine. Will look into it.

**Comment by TerrenceMcGuinness-NOAA** [comment]
Creating the experment **C48_S2SWA_gefs** failed on Hera on `KeyError: 'FHOUT_OCNICE`

```
Terry.McGuinness (hfe11) CI $ /scratch1/NCEPDEV/global/CI/2667/gefs/ci/scripts/utils/ci_utils_wrapper.sh create_experiment /scratch1/NCEPDEV/global/CI/2667/gefs/ci/cases/pr/C48_S2SWA_gefs.yaml

Running "module reset". Resetting modules to system default. The following $MODULEPATH directories have been removed: None
2024-06-11 15:40:28,849 - INFO     - root        : BEGIN: __main__.input_args
2024-06-11 15:40:28,849 - DEBUG    - root        : (  )
2024-06-11 15:40:28,850 - INFO     - root        :   END: __main__.input_args
2024-06-11 15:40:28,850 - DEBUG    - root        :  returning: Namespace(yaml=PosixPath('/scratch1/NCEPDEV/global/CI/2667/gefs/ci/cases/pr/C48_S2SWA_gefs.yaml'), overwrite=True)
2024-06-11 15:40:28,863 - INFO     - root        : Call: setup_expt.main()
2024-06-11 15:40:28,863 - DEBUG    - root        : setup_expt.py gefs forecast-only --pslot C48_S2SWA_gefs_6483983d --app S2SWA --resdetatmos 48 --resdetocean 5.0 --resensatmos 48 --nens 2 --gfs_cyc 1 --start cold --comroot UNDEFINED/COMROOT --expdir UNDEFINED/EXPDIR --idate 2021032312 --edate 2021032312 --yaml /scratch1/NCEPDEV/global/CI/2667/gefs/ci/cases/yamls/gefs_ci_defaults.yaml --overwrite
forecast-only mode treats ICs differently and cannot be staged here
EDITED:  UNDEFINED/EXPDIR/C48_S2SWA_gefs_6483983d/config.base as per user input.
****************************************************************************************************
EXPDIR: UNDEFINED/EXPDIR/C48_S2SWA_gefs_6483983d
ROTDIR: UNDEFINED/COMROOT/C48_S2SWA_gefs_6483983d
****************************************************************************************************
2024-06-11 15:40:29,019 - INFO     - root        : Call: setup_xml.main()
2024-06-11 15:40:29,019 - DEBUG    - root        : setup_xml.py /scratch1/NCEPDEV/global/CI/2667/gefs/UNDEFINED/EXPDIR/C48_S2SWA_gefs_6483983d
Finalizing initialize
sourcing config.stage_ic
sourcing config.fcst
sourcing config.atmos_products
sourcing config.efcs
sourcing config.atmos_ensstat
sourcing config.waveinit
sourcing config.wavepostsbs
sourcing config.wavepostpnt
sourcing config.oceanice_products
sourcing config.prep_emissions
component='atmos' fhmax=120 fhout=6
component='atmos' fhmax=120 fhout=6
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2667/gefs//workflow/create_experiment.py", line 101, in <module>
    setup_xml.main(setup_xml_args)
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/setup_xml.py", line 73, in main
    xml = rocoto_xml_factory.create(f'{net}_{mode}', app_config, rocoto_param_dict)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/wxflow/factory.py", line 71, in create
    return self._builders[key](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/gefs_xml.py", line 14, in __init__
    super().__init__(app_config, rocoto_config)
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/workflow_xml.py", line 27, in __init__
    task_list = get_wf_tasks(app_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/workflow_tasks.py", line 21, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/tasks.py", line 240, in get_task
    return getattr(self, task_name, *args, **kwargs)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/gefs_tasks.py", line 194, in ocean_prod
    return self._atmosoceaniceprod('ocean')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/gefs_tasks.py", line 250, in _atmosoceaniceprod
    fhrs = self._get_forecast_hours('gefs', self._configs[config], component)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/rocoto/tasks.py", line 136, in _get_forecast_hours
    local_config['FHOUT'] = config['FHOUT_OCNICE']
                            ~~~~~~^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2667/gefs/workflow/wxflow/attrdict.py", line 84, in __missing__
    raise KeyError(name)
KeyError: 'FHOUT_OCNICE'
Terry.McGuinness (hfe11) CI $ 
```

File: jobs/rocoto/atmos_products.sh [hunk 2]
```diff
@@ -15,21 +15,13 @@ if (( status != 0 )); then exit "${status}"; fi
 export job="atmos_products"
 export jobid="${job}.$$"
 
-###############################################################
-# shellcheck disable=SC2153,SC2001
-IFS='_' read -ra fhrs <<< "${FHRLST//f}" # strip off the 'f's and convert to array
+# Negatation needs to be before the base
+fhr3_base="10#${FHR3}"
+export FORECAST_HOUR=$(( ${fhr3_base/10#-/-10#} ))
```
Comment (DavidHuber-NOAA) [inline jobs/rocoto/atmos_products.sh:20]: The `#` in this regex breaks vim syntax highlighting unless it is escaped.
```suggestion
export FORECAST_HOUR=$(( ${fhr3_base/10\#-/-10\#} ))
```

**Review by DavidHuber-NOAA — APPROVED** [review]
Looks good, just one suggestion for vim users.

File: jobs/rocoto/atmos_products.sh (hunk 2 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/atmos_products.sh:20]: Not sure about making the code harder to read everywhere else to accommodate faulty parsers, even ones as widely used as `vim`.

**Comment by aerorahul** [comment]
> Experiment C96C48_hybatmDA **FAILED** on Hera with error logs:
> 
> ```
> /scratch1/NCEPDEV/global/CI/2667/RUNTESTS/COMROOT/C96C48_hybatmDA_a1ad6f4a/logs/2021122100/enkfgdasearc01.log
> ```
> 
> Follow link here to view the contents of the above file(s): [(link)](https://gist.github.com/emcbot/d54ef703bcc10c66f2b42180fba45519)

Source of this error is here.
https://github.com/NOAA-EMC/global-workflow/pull/2668/files#r1644852497

**Comment by aerorahul** [comment]
> Experiment C96C48_hybatmDA **FAILED** on Hera with error logs:
> 
> ```
> /scratch1/NCEPDEV/global/CI/2667/RUNTESTS/COMROOT/C96C48_hybatmDA_d5cf7918/logs/2021122100/enkfgdasearc01.log
> ```
> 
> Follow link here to view the contents of the above file(s): [(link)](https://gist.github.com/emcbot/aecc2b998d02217a1a0f0ebdabac2e40)

Ref. https://github.com/NOAA-EMC/global-workflow/pull/2668/files#r1646241550
Following the hotfix in #2697, the `COM_ATMOS_<>_TMPL` variables are also needed to resolve the jinja2 templates (in addition to the `COMIN`)
We need another hotfix for #2697

For some reason, I am not sure why the CI is not catching these.

**Pull request merged**